### PR TITLE
Allows non-strict load with distributed checkpoints

### PR DESCRIPTION
**Pull request by mikolajblaz**
# What does this PR do ?

With distributed checkpoints, the mismatches between the runtime model and checkpoint model manifest during `dist_checkpoint.load` (not during `model.load_state_dict` as with regular checkpoints).
This PR adds a flag that allows to adjust load strictness (e.g. ignore unexpected keys).

This PR relies on MCore feature that is not merged yet (merge ETA 5th July): https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/1628 

**Collection**: NLP

# Changelog 
- Add `model.dist_ckpt_load_strictness` flag to control dist ckpt load strictness. The most useful value is `log_all` which warns about all mismatches but performs the checkpoint load for a matching state dict subset.  

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: Dockerfile.ci
Comment (mikolajblaz) [inline Dockerfile.ci:56]: This change must be reverted once the MCore feature is merged

File: Dockerfile.ci
Comment (dimapihtar) [inline Dockerfile.ci:57]: can we wait until that changes merged to mcore?

Comment (mikolajblaz) [reply Dockerfile.ci:57]: yes

Comment (mikolajblaz) [reply Dockerfile.ci]: updated to public commit

**Review by dimapihtar — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**
